File: 2021/4/4.py
class Board:
    def __init__(self):
        self.rows = []
        self.cols = [set() for x in range(5)]

    def add_row(self, row: list[int]):
        self.rows.append(set(row))

        for i in range(len(row)):
            self.cols[i].add(row[i])

    def remove_guess(self, guess: int) -> bool:
        for row in self.rows:
            row.discard(guess)
            if not row:
                return True
        for col in self.cols:
            col.discard(guess)
            if not col:
                return True
        return False

    def sum_of_elements(self) -> int:
        return sum([sum(row) for row in self.rows])

    def print(self):
        for row in self.rows:
            print(row)

input_file = open("4.in")
guesses = [int(x) for x in input_file.readline().strip().split(',')]

boards = []
while input_file.readline():
    boards.append(Board())
    for i in range(5):
        row = [int(x) for x in input_file.readline().strip().split()]
        boards[-1].add_row(row)

input_file.close()

won = False
for guess in guesses:
    print(f"{guess = }")
    for board in boards:
        won = board.remove_guess(guess)
        # board.print()
        if won:
            print(board.sum_of_elements() * guess)
            break
    if won:
        break

# Task 2
input_file = open("4.in")
from collections import deque
guesses = deque([int(x) for x in input_file.readline().strip().split(',')])


boards = []
while input_file.readline():
    boards.append(Board())
    for i in range(5):
        row = [int(x) for x in input_file.readline().strip().split()]
        boards[-1].add_row(row)

input_file.close()

guess = 0
while len(boards) > 1:
    guess = guesses.popleft()
    boards = list(filter(lambda b: not b.remove_guess(guess), boards))

board = boards[0]
while True:
    guess = guesses.popleft()
    if board.remove_guess(guess):
        break

print(boards[0].sum_of_elements() * guess)

File: 2021/4/4.rb
class Board
  def initialize
    @rows = []
    @cols = Array.new(5) { Set.new }
  end

  def add_row(row)
    @rows << row.to_set
    row.each_with_index do |el, idx|
      @cols[idx] << el
    end
  end

  def remove_guess(guess)
    @rows.each { |r| r.delete(guess) }
    @cols.each { |c| c.delete(guess) }
  end

  def bingo?
    @rows.any?(&:empty?) or @cols.any?(&:empty?)
  end

  def sum_nums
    @rows.map(&:sum).sum
  end
end

input_file = File.open("4.in")
guesses = input_file.readline.chomp.split(",").map(&:to_i)
boards = []
while input_file.gets
  boards << Board.new
  5.times do
    row = input_file.readline.chomp.split.map(&:to_i)
    boards.last.add_row row
  end
end
input_file.close

guess = nil
done = false
until done
  guess = guesses.shift
  puts "Guess: #{guess}"
  boards.each do |board|
    board.remove_guess(guess)
    if board.bingo?
      puts board.sum_nums * guess
      done = true
      break
    end
  end
end


# Task 2
input_file = File.open("4.in")
guesses = input_file.readline.chomp.split(",").map(&:to_i)
boards = []
while input_file.gets
  boards << Board.new
  5.times do
    row = input_file.readline.chomp.split.map(&:to_i)
    boards.last.add_row row
  end
end
input_file.close


guess = nil
done = false
while boards.size > 1
  guess = guesses.shift
  puts "Guess: #{guess}"
  boards.each { |b| b.remove_guess(guess) }
  boards.reject!(&:bingo?)
end
board = boards.first
until board.bingo?
  guess = guesses.shift
  board.remove_guess(guess)
end
puts boards.first.sum_nums * guess
